### PR TITLE
miner: Not commit mining work of block before checkpoint

### PR DESCRIPTION
**Pull request by ucwong**
The mining work of block before checkpoint is not available. 
It is good for miner to avoid wasting pow

Workers will start committing new work when miner received downloader.done or downloader.failed event. It will happen from block 1 to latest block in some cases

### CODE REVIEW

**Comment by rjl493456442** [comment]
It's a good point. But I think we should somehow combine the original mechanism.

What the original mechanism does here is: 
- Monitor the status of the local chain
- No matter what kind of event we receive (from the downloader), exit the loop
- Start mining

There are several scenarios that can happen:
- We only sync a part of the chain(a failed event delivered)
- We sync all the blocks we need

For the case 1, we just waste the power for useless mining. So in my opinion, we should
do something for case 1 to ensure we actually sync the latest one.

Here are a few tricks we can apply:
- ensure the latest header beyond the checkpoint(if it's available)
- ensure the latest header is close enough by checking the header date.

**Comment by ucwong** [comment]
> It's a good point. But I think we should somehow combine the original mechanism.
> 
> What the original mechanism does here is:
> 
> * Monitor the status of the local chain
> * No matter what kind of event we receive (from the downloader), exit the loop
> * Start mining
> 
> There are several scenarios that can happen:
> 
> * We only sync a part of the chain(a failed event delivered)
> * We sync all the blocks from the peer, but they are not latest
> * We sync all the latest blocks
> 
> For the case 1 and case 2, we just waste the power for useless mining. So in my opinion, we should
> do something for case 1/2 to ensure we actually sync the latest one.
> 
> Here are a few tricks we can apply:
> 
> * ensure the latest header beyond the checkpoint(if it's available)
> * ensure the latest header is close enough by checking the header date.

Is it OK to add a new event to downloader, to notify the miner to start. We can check point and latest block in downloader to decide whether to start mining or not ?

**Comment by rjl493456442** [comment]
Here is an example https://github.com/ethereum/go-ethereum/blob/master/cmd/geth/main.go#L432

But I think we don't need to be very accurate. Whenever we receive a `Done` event and the header is close enough, then we can just start mining if it's required.

File: miner/worker.go [hunk 1]
```diff
@@ -860,6 +863,9 @@ func (w *worker) commitNewWork(interrupt *int32, noempty bool, timestamp int64)
 	}
 
 	num := parent.Number()
+	if num.Uint64() <= w.checkpoint {
```
Comment (rjl493456442) [inline miner/worker.go:866]: If it's in private network without any checkpoint, the checkpoint number here is 0.
If it's the first block to mine(the parent is genesis), the mining operation will be aborted.

**Comment by karalabe** [comment]
We've been discussing that instead of having th eminer be overly smart and start/stop itself, we should rather have the `eth` package direct this based on chain advancement. That requires a slight rework on the start/stop mechanism which is a bit racey currently. @MariusVanDerWijden and co will track that and we should pick this up afterwards.

**Comment by holiman** [comment]
After moving to PoS, this PR is no longer relevant, so I'm closing. Feel free to reopen if I have misunderstood. 
Thanks for the effort!

**Pull request closed without merging**